From david-b@pacbell.net  Tue Sep 23 16:02:44 2008
From: Yauhen Kharuzhy <jekhor@gmail.com>
Date: Fri, 12 Sep 2008 09:02:23 -0700
Subject: USB: Fix s3c2410_udc usb speed handling
To: Greg KH <greg@kroah.com>
Cc: linux-usb@vger.kernel.org, Ben Dooks <ben-linux@fluff.org>, Yauhen Kharuzhy <jekhor@gmail.com>
Message-ID: <200809120902.23903.david-b@pacbell.net>
Content-Disposition: inline

From: Yauhen Kharuzhy <jekhor@gmail.com>

The new composite framework revealed a weakness in the
s3c2410_udc driver gadget register function. Instead of
checking if speed asked for was USB_LOW_SPEED upon
usb_gadget_register() to deny service, it checked only
for USB_FULL_SPEED, thus denying service to usb high
speed capable gadgets (like g_ether).

Signed-off-by: Yauhen Kharuzhy <jekhor@gmail.com>
Cc: stable <stable@kernel.org>
Signed-off-by: David Brownell <dbrownell@users.sourceforge.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/s3c2410_udc.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/usb/gadget/s3c2410_udc.c
+++ b/drivers/usb/gadget/s3c2410_udc.c
@@ -1651,7 +1651,7 @@ int usb_gadget_register_driver(struct us
 		return -EBUSY;
 
 	if (!driver->bind || !driver->setup
-			|| driver->speed != USB_SPEED_FULL) {
+			|| driver->speed < USB_SPEED_FULL) {
 		printk(KERN_ERR "Invalid driver: bind %p setup %p speed %d\n",
 			driver->bind, driver->setup, driver->speed);
 		return -EINVAL;
